python/cudf/cudf/core/indexing_utils.py

```diff
@@ -0,0 +1,254 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Any, List, Tuple
+
+from typing_extensions import TypeAlias
+
+import cudf
+from cudf.api.types import (
+    _is_scalar_or_zero_d_array,
+    is_bool_dtype,
+    is_integer,
+    is_integer_dtype,
+)
+from cudf.core import copy_types as ct
+
+
+# Poor man's algebraic data types
+class EmptyIndexer:
+    """An indexer that will produce an empty result"""
+
+    pass
+
+
+@dataclass
+class MapIndexer:
+    """An indexer for a gather map"""
+
+    gather_map: ct.GatherMap
+
+
+@dataclass
+class MaskIndexer:
+    """An indexer for a boolean mask"""
+
+    mask: ct.BooleanMask
+
+
+@dataclass
+class SliceIndexer:
+    """An indexer for a slice"""
+
+    slice: slice
+
+
+@dataclass
+class ScalarIndexer:
+    """An indexer for a scalar value"""
+
+    gather_map: ct.GatherMap
+
+
+IndexingSpec: TypeAlias = (
+    EmptyIndexer | MapIndexer | MaskIndexer | ScalarIndexer | SliceIndexer
+)
+
+ColumnLabels: TypeAlias = List[str]
+
+
+def destructure_iloc_key(
+    key: Any, frame: cudf.Series | cudf.DataFrame
+) -> tuple[Any, ...]:
+    """
+    Destructure a potentially tuple-typed key into row and column indexers
+
+    Tuple arguments to iloc indexing are treated specially. They are
+    picked apart into indexers for the row and column. If the number
+    of entries is less than the number of modes of the frame, missing
+    entries are slice-expanded.
+
+    If the user-provided key is not a tuple, it is treated as if it
+    were a singleton tuple, and then slice-expanded.
+
+    Once this destructuring has occurred, any entries that are
+    callables are then called with the indexed frame. This should
+    return a valid indexing object for the rows (respectively
+    columns), namely one of:
+
+    - A boolean mask of the same length as the frame in the given
+      dimension
+    - A scalar integer that indexes the frame
+    - An array-like of integers that index the frame
+    - A slice that indexes the frame
+
+    Integer and slice-based indexing follows usual Python conventions.
+
+    Parameters
+    ----------
+    key
+        The key to destructure
+    frame
+        DataFrame or Series to provide context
+
+    Returns
+    -------
+    tuple of indexers with length equal to the dimension of the frame
+
+    Raises
+    ------
+    IndexError
+        If there are too many indexers, or any individual indexer is a tuple.
+    """
+    n = len(frame.shape)
+    if isinstance(key, tuple):
+        indexers = tuple(
+            itertools.chain(key, itertools.repeat(slice(None), n - len(key)))
```

Comment:
It's probably negligible in something of this complexity, but I have on occasion noted itertools as a significant perf bottleneck so it may be worth double-checking that it's not noticeable here.

Reply:
Did the "simple" thing since it's actually easier to read that way.